in_toto/user_settings.py

```diff
@@ -0,0 +1,132 @@
+    e.g. the envvar...
+    IN_TOTO_ARTIFACT_EXCLUDES="*.link:.gitignore"
+
+    becomes the setting ...
+    in_toto.settings.ARTIFACT_EXCLUDES = ["*.link", ".gitignore"]
+
+  - RCfile values follow the format expected by Python's builtin
+    ConfigParser. The format requires at least one section, sections however
+    are ignored, e.g. the rcfile...
+    [ignored section name]
+    artifact_basepath = "/a/b/c"
+
+    becomes the setting ...
+    in_toto.settings.ARTIFACT_BASEPATH = "/a/b/c"
+
+  - Order of precedence
+    1. RCfiles
+        .in_totorc
+        ~/.in_totorc
+        ~/.in_toto/config
+        ~/.config/in_toto
+        ~/.config/in_toto/config
+        /etc/in_totorc
+        /etc/in_toto/config
+    2. Environment Variables
+    3. Settings defined in `settings.py`
+
+"""
+import os
+import ConfigParser
+import in_toto.settings as SETTINGS_MODULE
+
+ENV_PREFIX = "IN_TOTO_"
+USER = os.path.expanduser("~")
+
+RC_PATHS = [
+  os.path.join("/etc", "in_toto", "config"),
+  os.path.join("/etc", "in_totorc"),
+  os.path.join(USER,  ".config", "in_toto", "config"),
+  os.path.join(USER, ".config", "in_toto"),
+  os.path.join(USER, ".in_toto", "config"),
+  os.path.join(USER, ".in_totorc"),
+  ".in_totorc"
+]
+
+def _colon_split(value):
+  """ If `value` contains colons, return a list split at colons,
+  return value otherwise. """
+  value_list = value.split(":")
+  if len(value_list) > 1:
+    return value_list
+
+  return value
+
+
+def get_env():
+  """Return a dict of environment variables starting with `ENV_PREFIX`
+
+  - In the returned dict the prefix is stripped from the keys
+  - Values that contain colons are converted into a list
+  - Variable names are converted to upper case
+
+  """
+  env_dict = {}
+
+  for name, value in os.environ.iteritems():
+    name = name.upper()
+
+    if (name.startswith(ENV_PREFIX) and
+        len(name) > len(ENV_PREFIX)):
+      stripped_name = name[len(ENV_PREFIX):]
+
+      env_dict[stripped_name] = _colon_split(value)
+
+  return env_dict
+
+
+def get_rc():
+  """Load RCfiles from the usual places and return a dictionary of
+  key value pairs.
+
+  - RCfile format as expected by builtin ConfigParser
+  - The format dictates that there is at least one section,
+    however we are only interested in the values and ignore the sections
+  - Values that contain colons are converted into a list
+  - Variable names are converted to upper case
+
+  """
+  rc_dict = {}
+
+  config = ConfigParser.ConfigParser()
+  config.read(RC_PATHS)
+
+  for section in config.sections():
+    for name, value in config.items(section):
+      name = name.upper()
+      rc_dict[name] = _colon_split(value)
+
+  return rc_dict
+
+
+def set_settings():
+  """Read environment variables and RCfiles and write them to the
+  settings module.
+  """
+  user_settings = get_env()
+  user_settings.update(get_rc())
+
+  for name, value in user_settings.iteritems():
+    setattr(SETTINGS_MODULE, name, value)
```

Comment:
Can you verify that someone can't cause unintended lines?  Is it possible to populate `in-toto.settings.py` with malicious input like: `"OS.REMOVE(my_files)".lower()`

Perhaps you don't have to worry about malicious input in config files and environment variables, but something to ponder.

Reply:
It does not seem possible, since I am assigning module variables with strings.

However, there isn't really a need to allow everything, so it might be better to make a whitelist, i.e. only allow settings for which there is a default in `settings.py`.

Reply:
Using a whitelist is the more secure approach in my opinion.